### PR TITLE
fix issue 21238 - -deps considers only the first instantiation site of a template for dependencies

### PR DESCRIPTION
**Pull request by aG0aep6G**
Note that this reopens [issue 9948](https://issues.dlang.org/show_bug.cgi?id=9948) which was considered "fixed" seven years ago. I believe that "fix" was faulty, and `-deps` has been in a broken, unusable state ever since, but there may be a way to resolve issue 21238 without reopening issue 9948. This PR does not attempt that, though.

### CODE REVIEW

File: test/compilable/depsOutput21238_foo.d [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+PERMUTE_ARGS:
+REQUIRED_ARGS: -deps -Jcompilable/imports
+EXTRA_SOURCES: imports/depsOutput21238_bar.d imports/depsOutput21238_baz_moduleimport.d imports/depsOutput21238_baz_stringimport.d imports/depsOutput21238_qux.d
+TRANSFORM_OUTPUT: remove_lines("druntime")
+TEST_OUTPUT:
+---
+depsImport depsOutput21238_foo (compilable$?:windows=\\|/$depsOutput21238_foo.d) : private : imports.depsOutput21238_baz_moduleimport (compilable$?:windows=\\|/$imports$?:windows=\\|/$depsOutput21238_baz_moduleimport.d)
```
Comment (wilzbach) [inline test/compilable/depsOutput21238_foo.d:8]: CC @MoonlightSentinel we should really find a nicer way than this :/

**Comment by ibuclaw** [comment]
Though genuine question, which tools still uses `-deps`? I thought I convinced the last consumer to get off it years ago.

**Comment by Geod24** [comment]
> Though genuine question, which tools still uses `-deps`? I thought I convinced the last consumer to get off it years ago.

[Well you should know](https://github.com/sociomantic-tsunami/makd/blob/89a4d5a72b2f99141c1c5a9a0ed110d3e56721ee/Makd.mak#L701).

**Comment by aG0aep6G** [comment]
> Though genuine question, which tools still uses `-deps`? I thought I convinced the last consumer to get off it years ago.

I don't know if there is any serious user. Given the broken state of `-deps`, I doubt it. My use case is just generating graph images. I can use dparse instead, but that gives me a full graph of possible imports, while `-deps` tells me what is actually being pulled in for a specific program. Is there another, truer alternative? I guess we could just ditch `-deps` in that case.

**Comment by ibuclaw** [comment]
> > Though genuine question, which tools still uses `-deps`? I thought I convinced the last consumer to get off it years ago.
> 
> I don't know if there is any serious user. Given the broken state of `-deps`, I doubt it. My use case is just generating graph images. I can use dparse instead, but that gives me a full graph of possible imports, while `-deps` tells me what is actually being pulled in for a specific program. Is there another, truer alternative? I guess we could just ditch `-deps` in that case.

It's coarse, but if you only care about dependencies at a file level (and not individual symbols), then `-v` prints all imports in a machine-ish format.

**Comment by ibuclaw** [comment]
> > Though genuine question, which tools still uses `-deps`? I thought I convinced the last consumer to get off it years ago.
> 
> [Well you should know](https://github.com/sociomantic-tsunami/makd/blob/89a4d5a72b2f99141c1c5a9a0ed110d3e56721ee/Makd.mak#L701).

It's not as if I spend my time reading tsunami code line by line. ;-)

**Comment by aG0aep6G** [comment]
> It's coarse, but if you only care about dependencies at a file level (and not individual symbols), then `-v` prints all imports in a machine-ish format.

`-v` is flat, though, isn't it? I.e., it tells me what is being pulled in, but it doesn't tell me where to draw the arrows that make the image interesting.

**Comment by ibuclaw** [comment]
> > It's coarse, but if you only care about dependencies at a file level (and not individual symbols), then `-v` prints all imports in a machine-ish format.
> 
> `-v` is flat, though, isn't it? I.e., it tells me what is being pulled in, but it doesn't tell me where to draw the arrows that make the image interesting.

Yes, it's flat, which is good enough if all you want to do is use it with a build system.  If you want to generate a dot file, you'd best use something else.

File: test/compilable/depsOutput21238_foo.d (hunk 1 above)
Comment (aG0aep6G) [reply test/compilable/depsOutput21238_foo.d]: How about `$p:<path>$`? I've pushed a commit.

Comment (aG0aep6G) [reply test/compilable/depsOutput21238_foo.d:8]: > How about `$p:<path>$`? I've pushed a commit.

Doesn't work. Apparently, Windows DMD prints double backslashes, but `$p:<path>$` expects singles. Back to `$?:windows=\\|/$`.

**Comment by aG0aep6G** [comment]
Ping?

**Comment by RazvanN7** [comment]
As far as I understand, the fix for Issue 9948 was actually an accidental one as the original PR actually had a different purpose. I think that @WalterBright is the only one who can actually make a judgement in this case.

**Comment by aG0aep6G** [comment]
> As far as I understand, the fix for Issue 9948 was actually an accidental one as the original PR actually had a different purpose.

I don't think that's accurate. There were two PRs: #2550 and #2553.

#2550 was Walter's, and it was unrelated to issue 9948. But the changes there apparently made fixing issue 9948 "trivial".

#2553 (not by Walter) is the one that fixed issue 9948, utilizing the changes from the previous PR.

I'm only reverting #2553 here. I'm not reverting anything from #2550. I'm not undoing any of Walter's work.

**Comment by aG0aep6G** [comment]
This has been idling for half a year now.

**Comment by aG0aep6G** [comment]
ping @RazvanN7 @WalterBright

This is good to go. It has been good to go from the very start. Just merge it already.

**Comment by thewilsonator** [comment]
Sorry for missing this. Could you please rebase, it seems that buildkite/druntime fails with something  rebase should fix.

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
Indeed, sorry for missing the notification.

**Comment by aG0aep6G** [comment]
> Could you please rebase, it seems that buildkite/druntime fails with something rebase should fix.

rebased

**Comment by thewilsonator** [comment]
thanks!

**Pull request merged**